lnwallet/channel_test.go

```diff
@@ -7812,3 +7811,385 @@ func TestFetchParent(t *testing.T) {
+				// Map their htlc settle update (3) to our htlc
+				// add (1).
+				map[uint64]uint64{3: 1},
+			),
+			expectedFee:      feePerKw,
+			ourExpectedHtlcs: nil,
+			theirExpectedHtlcs: map[uint64]bool{
+				2: true,
+			},
+			expectReceived: 0,
+			expectSent:     htlcAddAmount,
+		},
+		{
+			name:        "our htlc settled, state not mutated",
+			remoteChain: false,
+			mutateState: false,
+			ourHtlcs: []*PaymentDescriptor{
+				{
+					HtlcIndex: 1,
+					Amount:    htlcAddAmount,
+					EntryType: Add,
+				},
+			},
+			theirHtlcs: []*PaymentDescriptor{
+				{
+					HtlcIndex: 2,
+					Amount:    htlcAddAmount,
+					EntryType: Add,
+				},
+				{
+					HtlcIndex: 3,
+					Amount:    htlcAddAmount,
+					EntryType: Settle,
+				},
+			},
+			fetchParent: fetchParentFromMap(
+				// Map their htlc settle update (3) to our htlc
+				// add (1).
+				map[uint64]uint64{3: 1},
+			),
+			expectedFee:      feePerKw,
+			ourExpectedHtlcs: nil,
+			theirExpectedHtlcs: map[uint64]bool{
+				2: true,
+			},
+			expectReceived: 0,
+			expectSent:     0,
+		},
+		{
+			name:        "their htlc settled, state mutated",
+			remoteChain: false,
+			mutateState: true,
+			ourHtlcs: []*PaymentDescriptor{
+				{
+					HtlcIndex: 1,
+					Amount:    htlcAddAmount,
+					EntryType: Add,
+				},
+				{
+					HtlcIndex: 3,
+					Amount:    htlcAddAmount,
+					EntryType: Settle,
+				},
+			},
+			theirHtlcs: []*PaymentDescriptor{
+				{
+					HtlcIndex: 2,
+					Amount:    htlcAddAmount,
+					EntryType: Add,
+				},
+			},
+			fetchParent: fetchParentFromMap(
+				// Map our settle (3) to their add (2).
+				map[uint64]uint64{3: 2},
+			),
+			expectedFee: feePerKw,
+			ourExpectedHtlcs: map[uint64]bool{
+				1: true,
+			},
+			theirExpectedHtlcs: nil,
+			expectReceived:     htlcAddAmount,
+			expectSent:         0,
+		},
+		{
+			name:        "their htlc settled, state not mutated",
+			remoteChain: false,
+			mutateState: false,
+			ourHtlcs: []*PaymentDescriptor{
+				{
+					HtlcIndex: 1,
+					Amount:    htlcAddAmount,
+					EntryType: Add,
+				},
+				{
+					HtlcIndex: 3,
+					Amount:    htlcAddAmount,
+					EntryType: Settle,
+				},
+			},
+			theirHtlcs: []*PaymentDescriptor{
+				{
+					HtlcIndex: 2,
+					Amount:    htlcAddAmount,
+					EntryType: Add,
+				},
+			},
+			fetchParent: fetchParentFromMap(
+				// Map our settle (3) to their add (2).
+				map[uint64]uint64{3: 2},
+			),
+			expectedFee: feePerKw,
+			ourExpectedHtlcs: map[uint64]bool{
+				1: true,
+			},
+			theirExpectedHtlcs: nil,
+			expectReceived:     0,
+			expectSent:         0,
+		},
+	}
+
+	for _, test := range tests {
+		test := test
+
+		t.Run(test.name, func(t *testing.T) {
+			lc := LightningChannel{
+				channelState: &channeldb.OpenChannel{
+					TotalMSatSent:     0,
+					TotalMSatReceived: 0,
+				},
+			}
+
+			view := &htlcView{
+				ourUpdates:   test.ourHtlcs,
+				theirUpdates: test.theirHtlcs,
+				feePerKw:     feePerKw,
+			}
```

Comment:
I am wondering, is this indeed the right abstraction level for this test? I can see that the goal is to test `evaluteHtlcView` in isolation, but perhaps it is better to test at a slightly higher level to be sure that the interaction between `fetchHTLCView` and `evaluateHtlcView` is captured as well.

Reply:
I think there's too much going on in `evaluateHtlcView` to achieve this level of coverage at a higher level. There's a lot of subtlety with state mutation, which log we use etc. If anything, another test for `computeView` would be nice.